In [9]:
import os

project_files = [
    "data",
    "notebooks",
    "app.py",
    "requirements.txt",
    "README.md"
]

for item in project_files:
    print(item, "=>", os.path.exists(item))

data => True
notebooks => True
app.py => True
requirements.txt => True
README.md => True


In [5]:
import os

print("Current working directory:")
print(os.getcwd())

print("\nFiles in this place:")
print(os.listdir())

Current working directory:
y:\Courses\DataScience\EpsilonAI\final-project\BRFSS-Diabetes-Prediction

Files in this place:
['.ipynb_checkpoints', 'app.py', 'data', 'notebooks', 'README.md', 'requirements.txt']


In [11]:
import os

print(os.listdir("data"))

['LLCP2024.XPT']


In [12]:
import pandas as pd

data_path = "data/LLCP2024.XPT"

data = pd.read_sas(data_path, format="xport", encoding="latin1")

print("Data loaded successfully!")
print("Shape:", data.shape)

Data loaded successfully!
Shape: (457670, 301)


In [13]:
data.head()

,_STATE,FMONTH,IDATE,IMONTH,IDAY,IYEAR,DISPCODE,SEQNO,_PSU,CTELENM1,...,_LCSCTSN,_LCSPSTF,DRNKANY6,DROCDY4_,_RFBING6,_DRNKWK3,_RFDRHV9,_FLSHOT7,_PNEUMO3,_AIDTST4
0,1.0,2.0,02282024,02,28,2024,1100.0,2024000001,2.024000e+09,1.0,...,NaN,9.0,2.0,5.397605e-79,1.0,5.397605e-79,1.0,1.0,2.0,2.0
1,1.0,2.0,02212024,02,21,2024,1100.0,2024000002,2.024000e+09,1.0,...,4.0,9.0,2.0,5.397605e-79,1.0,5.397605e-79,1.0,1.0,1.0,2.0
2,1.0,2.0,02212024,02,21,2024,1100.0,2024000003,2.024000e+09,1.0,...,4.0,2.0,1.0,1.000000e+02,2.0,1.400000e+03,1.0,NaN,NaN,2.0
3,1.0,2.0,02282024,02,28,2024,1100.0,2024000004,2.024000e+09,1.0,...,NaN,9.0,2.0,5.397605e-79,1.0,5.397605e-79,1.0,1.0,1.0,2.0
4,1.0,2.0,02212024,02,21,2024,1100.0,2024000005,2.024000e+09,1.0,...,3.0,9.0,2.0,5.397605e-79,1.0,5.397605e-79,1.0,NaN,NaN,2.0


In [14]:
print(data.columns.tolist())

['_STATE', 'FMONTH', 'IDATE', 'IMONTH', 'IDAY', 'IYEAR', 'DISPCODE', 'SEQNO', '_PSU', 'CTELENM1', 'PVTRESD1', 'COLGHOUS', 'STATERE1', 'CELPHON1', 'LADULT1', 'NUMADULT', 'RESPSLC1', 'LANDSEX3', 'SAFETIME', 'CTELNUM1', 'CELLFON5', 'CADULT1', 'CELLSEX3', 'PVTRESD3', 'CCLGHOUS', 'CSTATE1', 'LANDLINE', 'HHADULT', 'SEXVAR', 'GENHLTH', 'PHYSHLTH', 'MENTHLTH', 'POORHLTH', 'PRIMINS2', 'PERSDOC3', 'MEDCOST1', 'CHECKUP1', 'EXERANY2', 'LASTDEN4', 'RMVTETH4', 'CVDINFR4', 'CVDCRHD4', 'CVDSTRK3', 'ASTHMA3', 'ASTHNOW', 'CHCSCNC1', 'CHCOCNC1', 'CHCCOPD3', 'ADDEPEV3', 'CHCKDNY2', 'HAVARTH4', 'DIABETE4', 'DIABAGE4', 'MARITAL', 'EDUCA', 'RENTHOM1', 'NUMHHOL4', 'NUMPHON4', 'CPDEMO1C', 'VETERAN3', 'EMPLOY1', 'CHILDREN', 'INCOME3', 'PREGNANT', 'WEIGHT2', 'HEIGHT3', 'DEAF', 'BLIND', 'DECIDE', 'DIFFWALK', 'DIFFDRES', 'DIFFALON', 'HADMAM', 'HOWLONG', 'CERVSCRN', 'CRVCLCNC', 'CRVCLPAP', 'CRVCLHPV', 'HADHYST2', 'HADSIGM4', 'COLNSIGM', 'COLNTES1', 'SIGMTES1', 'LASTSIG4', 'COLNCNCR', 'VIRCOLO1', 'VCLNTES2', 'SMALST

In [16]:
selected_columns = [
    "DIABETE4",    
    "_BMI5",
    "_RFHLTH",
    "PHYSHLTH",
    "MENTHLTH",
    "CVDSTRK3",
    "_MICHD",
    "_SMOKER3",
    "_TOTINDA",
    "SEXVAR",
    "_AGEG5YR",
    "_EDUCAG",
    "_INCOMG1",
    "_RFDRHV9"
]

missing_columns = []

for col in selected_columns:
    if col not in data.columns:
        missing_columns.append(col)

print("Number of selected columns:", len(selected_columns))
print("Missing columns:", missing_columns)

Number of selected columns: 14
Missing columns: []


In [17]:
df = data[selected_columns].copy()

print("Original data shape:", data.shape)
print("Selected data shape:", df.shape)

Original data shape: (457670, 301)
Selected data shape: (457670, 14)


In [18]:
df.head()

,DIABETE4,_BMI5,_RFHLTH,PHYSHLTH,MENTHLTH,CVDSTRK3,_MICHD,_SMOKER3,_TOTINDA,SEXVAR,_AGEG5YR,_EDUCAG,_INCOMG1,_RFDRHV9
0,3.0,2249.0,1.0,2.0,88.0,2.0,2.0,4.0,1.0,2.0,12.0,2.0,9.0,1.0
1,3.0,2583.0,1.0,88.0,88.0,2.0,1.0,3.0,1.0,1.0,13.0,4.0,7.0,1.0
2,3.0,2253.0,1.0,30.0,88.0,2.0,2.0,1.0,1.0,1.0,8.0,3.0,9.0,1.0
3,3.0,2509.0,1.0,88.0,88.0,2.0,2.0,4.0,1.0,1.0,13.0,4.0,4.0,1.0
4,3.0,1977.0,1.0,88.0,88.0,2.0,2.0,4.0,2.0,1.0,6.0,3.0,2.0,1.0


In [19]:
df.columns.tolist()

['DIABETE4',
 '_BMI5',
 '_RFHLTH',
 'PHYSHLTH',
 'MENTHLTH',
 'CVDSTRK3',
 '_MICHD',
 '_SMOKER3',
 '_TOTINDA',
 'SEXVAR',
 '_AGEG5YR',
 '_EDUCAG',
 '_INCOMG1',
 '_RFDRHV9']

In [20]:
df.rename(columns={
    "DIABETE4": "Diabetes_012",
    "_BMI5": "BMI",
    "_RFHLTH": "GeneralHealth",
    "PHYSHLTH": "PhysicalHealthDays",
    "MENTHLTH": "MentalHealthDays",
    "CVDSTRK3": "Stroke",
    "_MICHD": "HeartDisease",
    "_SMOKER3": "Smoker",
    "_TOTINDA": "PhysicalActivity",
    "SEXVAR": "Sex",
    "_AGEG5YR": "Age",
    "_EDUCAG": "Education",
    "_INCOMG1": "Income",
    "_RFDRHV9": "HeavyAlcohol"
}, inplace=True)

df.head()

,Diabetes_012,BMI,GeneralHealth,PhysicalHealthDays,MentalHealthDays,Stroke,HeartDisease,Smoker,PhysicalActivity,Sex,Age,Education,Income,HeavyAlcohol
0,3.0,2249.0,1.0,2.0,88.0,2.0,2.0,4.0,1.0,2.0,12.0,2.0,9.0,1.0
1,3.0,2583.0,1.0,88.0,88.0,2.0,1.0,3.0,1.0,1.0,13.0,4.0,7.0,1.0
2,3.0,2253.0,1.0,30.0,88.0,2.0,2.0,1.0,1.0,1.0,8.0,3.0,9.0,1.0
3,3.0,2509.0,1.0,88.0,88.0,2.0,2.0,4.0,1.0,1.0,13.0,4.0,4.0,1.0
4,3.0,1977.0,1.0,88.0,88.0,2.0,2.0,4.0,2.0,1.0,6.0,3.0,2.0,1.0


In [21]:
df.columns.tolist()

['Diabetes_012',
 'BMI',
 'GeneralHealth',
 'PhysicalHealthDays',
 'MentalHealthDays',
 'Stroke',
 'HeartDisease',
 'Smoker',
 'PhysicalActivity',
 'Sex',
 'Age',
 'Education',
 'Income',
 'HeavyAlcohol']

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 457670 entries, 0 to 457669
Data columns (total 14 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   Diabetes_012        457666 non-null  float64
 1   BMI                 414633 non-null  float64
 2   GeneralHealth       457670 non-null  float64
 3   PhysicalHealthDays  457665 non-null  float64
 4   MentalHealthDays    457667 non-null  float64
 5   Stroke              457667 non-null  float64
 6   HeartDisease        452464 non-null  float64
 7   Smoker              457670 non-null  float64
 8   PhysicalActivity    457670 non-null  float64
 9   Sex                 457670 non-null  float64
 10  Age                 457670 non-null  float64
 11  Education           457670 non-null  float64
 12  Income              457670 non-null  float64
 13  HeavyAlcohol        457670 non-null  float64
dtypes: float64(14)
memory usage: 48.9 MB


In [23]:
df.isnull().sum()

Diabetes_012              4
BMI                   43037
GeneralHealth             0
PhysicalHealthDays        5
MentalHealthDays          3
Stroke                    3
HeartDisease           5206
Smoker                    0
PhysicalActivity          0
Sex                       0
Age                       0
Education                 0
Income                    0
HeavyAlcohol              0
dtype: int64

In [24]:
for col in df.columns:
    print("=" * 50)
    print(col)
    print(df[col].value_counts(dropna=False).sort_index())

Diabetes_012
Diabetes_012
1.0     65809
2.0      3395
3.0    376125
4.0     11307
7.0       798
9.0       232
NaN         4
Name: count, dtype: int64
BMI
BMI
1200.0        1
1205.0        2
1208.0        1
1211.0        5
1213.0        1
          ...  
9765.0        1
9844.0        1
9979.0        1
9984.0        1
NaN       43037
Name: count, Length: 4199, dtype: int64
GeneralHealth
GeneralHealth
1.0    366237
2.0     90123
9.0      1310
Name: count, dtype: int64
PhysicalHealthDays
PhysicalHealthDays
1.0      18505
2.0      27903
3.0      16919
4.0       9023
5.0      16155
6.0       2495
7.0       9385
8.0       1856
9.0        418
10.0     11130
11.0       135
12.0      1120
13.0       156
14.0      4667
15.0      9805
16.0       299
17.0       198
18.0       343
19.0        57
20.0      6492
21.0       982
22.0       209
23.0       117
24.0       118
25.0      2494
26.0       139
27.0       246
28.0       897
29.0       411
30.0     36353
77.0      8765
88.0    267576
99.0      22

In [25]:
df_clean = df.copy()

print("Original df shape:", df.shape)
print("Cleaning copy shape:", df_clean.shape)

Original df shape: (457670, 14)
Cleaning copy shape: (457670, 14)


 و احنا احطرت بين multi-class: binaray 
 عتبرنا pre-diabetes كـ non-diabetic class في binary classification.

In [26]:
# Keep only valid target values for binary classification
df_clean = df_clean[df_clean["Diabetes_012"].isin([1.0, 3.0, 4.0])]

# Map target values:
# 1.0 = Diabetes
# 3.0 = No Diabetes
# 4.0 = Pre-diabetes / borderline diabetes, treated as No Diabetes
df_clean["Diabetes_012"] = df_clean["Diabetes_012"].replace({
    1.0: 1,
    3.0: 0,
    4.0: 0
})

print("Shape after cleaning target:", df_clean.shape)
print(df_clean["Diabetes_012"].value_counts())

Shape after cleaning target: (453241, 14)
Diabetes_012
0.0    387432
1.0     65809
Name: count, dtype: int64


In [27]:
print("BMI missing before cleaning:")
print(df_clean["BMI"].isnull().sum())

print("\nBMI summary before scaling:")
print(df_clean["BMI"].describe())

BMI missing before cleaning:
42248

BMI summary before scaling:
count    410993.000000
mean       2854.774957
std         657.578042
min        1200.000000
25%        2414.000000
50%        2744.000000
75%        3175.000000
max        9984.000000
Name: BMI, dtype: float64


In [28]:
# BRFSS stores BMI as BMI * 100, so we divide by 100
df_clean["BMI"] = df_clean["BMI"] / 100

print("BMI missing after scaling:")
print(df_clean["BMI"].isnull().sum())

print("\nBMI summary after scaling:")
print(df_clean["BMI"].describe())

BMI missing after scaling:
42248

BMI summary after scaling:
count    410993.00000
mean         28.54775
std           6.57578
min          12.00000
25%          24.14000
50%          27.44000
75%          31.75000
max          99.84000
Name: BMI, dtype: float64


عندنا لسه Missing في BMI:

BMI missing = 42248

مش هنملاه دلوقتي يدويًا، لأننا لاحقًا في الـ ML Pipeline هنستخدم SimpleImputer عشان يعالج الـ missing values بطريقة منظمة داخل الـ preprocessing pipeline.

In [29]:
import numpy as np

health_days_cols = ["PhysicalHealthDays", "MentalHealthDays"]

for col in health_days_cols:
    df_clean[col] = df_clean[col].replace({
        88.0: 0,
        77.0: np.nan,
        99.0: np.nan
    })

print("PhysicalHealthDays value counts:")
print(df_clean["PhysicalHealthDays"].value_counts(dropna=False).sort_index())

print("\nMentalHealthDays value counts:")
print(df_clean["MentalHealthDays"].value_counts(dropna=False).sort_index())

PhysicalHealthDays value counts:
PhysicalHealthDays
0.0     265321
1.0      18334
2.0      27596
3.0      16753
4.0       8931
5.0      15974
6.0       2458
7.0       9271
8.0       1833
9.0        413
10.0     10991
11.0       132
12.0      1110
13.0       151
14.0      4616
15.0      9702
16.0       297
17.0       195
18.0       338
19.0        54
20.0      6415
21.0       960
22.0       203
23.0       116
24.0       118
25.0      2468
26.0       137
27.0       239
28.0       879
29.0       405
30.0     35988
NaN      10843
Name: count, dtype: int64

MentalHealthDays value counts:
MentalHealthDays
0.0     267874
1.0      15258
2.0      24903
3.0      15780
4.0       8274
5.0      21114
6.0       2336
7.0       8185
8.0       1850
9.0        331
10.0     16261
11.0       144
12.0      1326
13.0       185
14.0      2802
15.0     15102
16.0       296
17.0       287
18.0       343
19.0        70
20.0      9580
21.0       544
22.0       185
23.0       131
24.0       131
25.0      3305
26.

In [30]:
binary_cols = [
    "GeneralHealth",
    "Stroke",
    "HeartDisease",
    "PhysicalActivity",
    "HeavyAlcohol"
]

In [31]:
# GeneralHealth: 1 = Good/Better health, 2 = Fair/Poor health, 9 = Unknown
df_clean["GeneralHealth"] = df_clean["GeneralHealth"].replace({
    1.0: 1,
    2.0: 0,
    9.0: np.nan
})

# Stroke: 1 = Yes, 2 = No, 7/9 = Unknown/Refused
df_clean["Stroke"] = df_clean["Stroke"].replace({
    1.0: 1,
    2.0: 0,
    7.0: np.nan,
    9.0: np.nan
})

# HeartDisease: 1 = Yes, 2 = No
df_clean["HeartDisease"] = df_clean["HeartDisease"].replace({
    1.0: 1,
    2.0: 0
})

# PhysicalActivity: 1 = Yes, 2 = No, 9 = Unknown
df_clean["PhysicalActivity"] = df_clean["PhysicalActivity"].replace({
    1.0: 1,
    2.0: 0,
    9.0: np.nan
})

# HeavyAlcohol: 1 = Not heavy drinking, 2 = Heavy drinking, 9 = Unknown
df_clean["HeavyAlcohol"] = df_clean["HeavyAlcohol"].replace({
    1.0: 0,
    2.0: 1,
    9.0: np.nan
})

In [32]:
check_cols = [
    "GeneralHealth",
    "Stroke",
    "HeartDisease",
    "PhysicalActivity",
    "HeavyAlcohol"
]

for col in check_cols:
    print("=" * 40)
    print(col)
    print(df_clean[col].value_counts(dropna=False).sort_index())

GeneralHealth
GeneralHealth
0.0     89137
1.0    362862
NaN      1242
Name: count, dtype: int64
Stroke
Stroke
0.0    431523
1.0     20473
NaN      1245
Name: count, dtype: int64
HeartDisease
HeartDisease
0.0    406256
1.0     42072
NaN      4913
Name: count, dtype: int64
PhysicalActivity
PhysicalActivity
0.0    105159
1.0    346848
NaN      1234
Name: count, dtype: int64
HeavyAlcohol
HeavyAlcohol
0.0    383188
1.0     23954
NaN     46099
Name: count, dtype: int64


In [33]:
df_clean["Smoker"] = df_clean["Smoker"].replace({
    1.0: 1,
    2.0: 1,
    3.0: 1,
    4.0: 0,
    9.0: np.nan
})

In [38]:
print(df_clean["Smoker"].value_counts(dropna=False).sort_index())

Smoker
0.0    256434
1.0    165243
NaN     31564
Name: count, dtype: int64


In [39]:
df_clean["Sex"] = df_clean["Sex"].replace({
    1.0: 1,
    2.0: 0
})

In [40]:
print(df_clean["Sex"].value_counts(dropna=False).sort_index())

Sex
0.0    236324
1.0    216917
Name: count, dtype: int64


In [41]:
ordinal_cols = ["Age", "Education", "Income"]

for col in ordinal_cols:
    print("=" * 40)
    print(col)
    print(df_clean[col].value_counts(dropna=False).sort_index())

Age
Age
1.0     29469
2.0     23443
3.0     25781
4.0     28286
5.0     30355
6.0     28598
7.0     31376
8.0     34619
9.0     43086
10.0    47392
11.0    44556
12.0    36652
13.0    41537
14.0     8091
Name: count, dtype: int64
Education
Education
1.0     26740
2.0    114805
3.0    119613
4.0    189843
9.0      2240
Name: count, dtype: int64
Income
Income
1.0     20524
2.0     33077
3.0     41302
4.0     49259
5.0    111561
6.0     81748
7.0     29442
9.0     86328
Name: count, dtype: int64


In [42]:
df_clean["Education"] = df_clean["Education"].replace({
    9.0: np.nan
})

df_clean["Income"] = df_clean["Income"].replace({
    9.0: np.nan
})

In [43]:
for col in ["Age", "Education", "Income"]:
    print("=" * 40)
    print(col)
    print(df_clean[col].value_counts(dropna=False).sort_index())

Age
Age
1.0     29469
2.0     23443
3.0     25781
4.0     28286
5.0     30355
6.0     28598
7.0     31376
8.0     34619
9.0     43086
10.0    47392
11.0    44556
12.0    36652
13.0    41537
14.0     8091
Name: count, dtype: int64
Education
Education
1.0     26740
2.0    114805
3.0    119613
4.0    189843
NaN      2240
Name: count, dtype: int64
Income
Income
1.0     20524
2.0     33077
3.0     41302
4.0     49259
5.0    111561
6.0     81748
7.0     29442
NaN     86328
Name: count, dtype: int64


In [44]:
for col in df_clean.columns:
    print("=" * 50)
    print(col)
    print(df_clean[col].value_counts(dropna=False).sort_index())

Diabetes_012
Diabetes_012
0.0    387432
1.0     65809
Name: count, dtype: int64
BMI
BMI
12.00        1
12.05        2
12.08        1
12.11        5
12.13        1
         ...  
97.65        1
98.44        1
99.79        1
99.84        1
NaN      42248
Name: count, Length: 4186, dtype: int64
GeneralHealth
GeneralHealth
0.0     89137
1.0    362862
NaN      1242
Name: count, dtype: int64
PhysicalHealthDays
PhysicalHealthDays
0.0     265321
1.0      18334
2.0      27596
3.0      16753
4.0       8931
5.0      15974
6.0       2458
7.0       9271
8.0       1833
9.0        413
10.0     10991
11.0       132
12.0      1110
13.0       151
14.0      4616
15.0      9702
16.0       297
17.0       195
18.0       338
19.0        54
20.0      6415
21.0       960
22.0       203
23.0       116
24.0       118
25.0      2468
26.0       137
27.0       239
28.0       879
29.0       405
30.0     35988
NaN      10843
Name: count, dtype: int64
MentalHealthDays
MentalHealthDays
0.0     267874
1.0      15258
2.0

In [45]:
missing_summary = df_clean.isnull().sum().sort_values(ascending=False)

print(missing_summary)

Income                86328
HeavyAlcohol          46099
BMI                   42248
Smoker                31564
PhysicalHealthDays    10843
MentalHealthDays       7968
HeartDisease           4913
Education              2240
Stroke                 1245
GeneralHealth          1242
PhysicalActivity       1234
Diabetes_012              0
Sex                       0
Age                       0
dtype: int64


In [46]:
cleaned_data_path = "data/brfss_diabetes_cleaned.csv"

df_clean.to_csv(cleaned_data_path, index=False)

print("Cleaned dataset saved successfully!")
print("Path:", cleaned_data_path)
print("Shape:", df_clean.shape)

Cleaned dataset saved successfully!
Path: data/brfss_diabetes_cleaned.csv
Shape: (453241, 14)


In [47]:
import os

print(os.path.exists("data/brfss_diabetes_cleaned.csv"))

True


## Data Cleaning Summary

In this step, I cleaned the selected BRFSS 2024 dataset columns to prepare them for analysis and machine learning.

The target column `Diabetes_012` was converted into a binary classification target:
- `1` = Diabetes
- `0` = No Diabetes

Rows with unclear diabetes responses such as "Don't know", "Refused", missing values, and diabetes only during pregnancy were removed.

The `BMI` column was divided by 100 because BRFSS stores BMI values as BMI × 100.

For `PhysicalHealthDays` and `MentalHealthDays`:
- `88` was converted to `0`, meaning no unhealthy days.
- `77` and `99` were converted to missing values because they represent unknown or refused answers.

For binary health-related columns such as `Stroke`, `HeartDisease`, `PhysicalActivity`, `GeneralHealth`, and `HeavyAlcohol`, valid answers were mapped to `0` and `1`, while unknown or refused values were converted to missing values.

For `Smoker`, current smokers and former smokers were grouped as `1`, while never smokers were mapped to `0`.

For `Sex`, male was mapped to `1` and female was mapped to `0`.

For `Education` and `Income`, unknown values coded as `9` were converted to missing values.

Missing values were not dropped at this stage because they will be handled later using a preprocessing pipeline with imputation.